Applications/Spire/Source/Ui/PercentBox.cpp

```diff
@@ -0,0 +1,136 @@
+#include "Spire/Ui/PercentBox.hpp"
+#include "Spire/Spire/LocalScalarValueModel.hpp"
+#include "Spire/Ui/Layouts.hpp"
+
+using namespace boost;
+using namespace boost::signals2;
+using namespace Spire;
+using namespace Spire::Styles;
+
+namespace {
+  auto to_percentage(const optional<Decimal>& value) {
+    if(value) {
+      return optional<Decimal>(*value * 100);
+    }
+    return value;
+  }
+
+  auto to_quantity(const optional<Decimal>& value) {
+    if(value) {
+      return optional<Decimal>(*value / 100);
+    }
+    return value;
+  }
+
+  struct OptionalPercentModel : OptionalDecimalModel {
+    mutable UpdateSignal m_update_signal;
+    std::shared_ptr<OptionalDecimalModel> m_model;
+    optional<Decimal> m_value;
+    scoped_connection m_connection;
+
+    OptionalPercentModel(std::shared_ptr<OptionalDecimalModel> model)
+      : m_model(std::move(model)),
+        m_value(m_model->get()),
+        m_connection(m_model->connect_update_signal(
+          std::bind_front(&OptionalPercentModel::on_update, this))) {}
+
+    QValidator::State get_state() const {
+      return m_model->get_state();
+    }
+
+    optional<Decimal> get_minimum() const {
+      return to_percentage(m_model->get_minimum());
+    }
+
+    optional<Decimal> get_maximum() const {
+      return to_percentage(m_model->get_maximum());
+    }
+
+    Decimal get_increment() const {
+      return m_model->get_increment();
+    }
+
+    const optional<Decimal>& get() const override {
+      return m_value;
+    }
+
+    QValidator::State test(const optional<Decimal>& value) const override {
+      return m_model->test(to_quantity(value));
+    }
+
+    QValidator::State set(const optional<Decimal>& value) override {
+      m_value = value;
+      return m_model->set(to_quantity(value));
+    }
+
+    connection connect_update_signal(
+      const UpdateSignal::slot_type& slot) const override {
+      return m_update_signal.connect(slot);
+    }
+
+    void on_update(const optional<Decimal>& value) {
+      m_value = to_percentage(value);
+      m_update_signal(m_value);
+    }
+  };
+}
+
+PercentBox::PercentBox(QWidget* parent)
+  : PercentBox(QHash<Qt::KeyboardModifier, Decimal>(
+      {{Qt::NoModifier, 1}, {Qt::AltModifier, 5}, {Qt::ControlModifier, 10},
+      {Qt::ShiftModifier, 20}}), parent) {}
+
+PercentBox::PercentBox(
+    QHash<Qt::KeyboardModifier, Decimal> modifiers, QWidget* parent)
+  : PercentBox(std::make_shared<LocalOptionalDecimalModel>(),
+    std::move(modifiers), parent) {}
+
+PercentBox::PercentBox(std::shared_ptr<OptionalDecimalModel> model,
+    QHash<Qt::KeyboardModifier, Decimal> modifiers, QWidget* parent)
+    : QWidget(parent),
+      m_model(std::move(model)) {
+  m_decimal_box = new DecimalBox(
+    std::make_shared<OptionalPercentModel>(m_model), std::move(modifiers));
+  enclose(*this, *m_decimal_box);
+  setFocusProxy(m_decimal_box);
+  proxy_style(*this, *m_decimal_box);
+  update_style(*m_decimal_box, [] (auto& style) {});
```

Comment:
I believe line 97 can be removed.

Reply:
Here line 97 is used to resolve the issue that setting the `TextAlign` of the `PercentBox` to `AlignRight` makes the `DecimalBox` lose the `PaddingRight` style. 
```
      update_style(*PercentBox, [&] (auto& style) {
        style.get(Any()).set(TextAlign(Qt::AlignRight));
      });
```
The update style code above invalidates the style settings in the DecimalBox below:
```
    style.get(+(Any() > is_a<TextBox>()) %
        (is_a<Button>() && matches(Visibility::VISIBLE))).
      set(PaddingRight(scale_width(24)));
```
It is just a workaround. I think there might be something wrong with the styling for the DecimalBox.
Do you know how to style the TextBox in the DecimalBox with the PaddingRight of is 24px when buttons in the DecimalBox are visible? I've spent a long time trying to figure it out, but failed.

Reply:
I will remove this line and create another task for this issue.